docs/design/proposals/codegen-refactor/centralize_1.md

```diff
@@ -0,0 +1,197 @@
+
+```
+
+3. Look for other wrappers and access to `ackgenconfig` and consildate to `config` package
+
+```
+// Current
+func getSortedLateInitFieldsAndConfig(cfg *ackgenconfig.Config, r *model.CRD) ([]string, map[string]*ackgenconfig.LateInitializeConfig)
+
+
+// Proposed
+func (c *Config) GetLateInitFieldsAndConfig(cfg *ackgenconfig.Config, r *model.CRD) (map[string]*ackgenconfig.LateInitializeConfig, error)
+
+// Note, this simply gets the config; sorting and any processing AFTER getting config is done by the client
+
+```
+
+#### Centralize `reconciliation`
+Similar to the above, move `reconcialiation`-related functions from `code` package to `ackmodel` to realign responsibilities.
+
+1. Refactor `type CRD struct` in [crd.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/model/crd.go#L64) to remove `sdkapi`  as attribute and fix breakage
+  * move methods associated with attributes to `ackmodel`
+
+```
+// Current
+func (r *CRD) TypeRenames() map[string]string
+
+//Proposed
+func (m *Model) GetTypeRenames() (map[string]string, error)
+
+```
+
+2. Move code in `pkg/code/` related to `reconciliation` to `pkg/model/`
+  * previous dependencies on invalid helpers (ex: `r.GetAllRenames`) should be resolved
+  * ex: most/all of [common.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/generate/code/common.go) has nothing to do with generating code; it retrieves information related to `aws-sdk` and `ackgenconfig` regarding `shapes`, `ops`
+
+```
+// Current
+func FindIdentifiersInShape(r *model.CRD, shape *awssdkmodel.Shape, op *awssdkmodel.Operation) []string
+
+// Proposed
+func (m *Model) GetIdentifiersInShape(r *model.CRD, shape *awssdkmodel.Shape, op *awssdkmodel.Operation) ([]string, error)
+
+```
+
+* Note, for scenarios that do both `reconcile` and output Go code (ex: [late_initialize::FindLateInitializedFieldNames](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/generate/code/late_initialize.go#L31), these will need to be refactored into separate pieces prior to migration
+
+
+#### Align access/interface for `__ForStruct` funcs
+The goal is to make it easier for clients to use `SetResource` and `SetSDK` by removing the need to know the target type.
+
+1. Refactor access (public vs. private) for `SetResourceFor___`, and `SetSDKFor___`
+  * `setResourceForContainer` and `setSDKForContainer` are private while `SetResourceForStruct` and `SetSDKForStruct` are public despite the former pair being more general than the latter. Access should be swapped so users calling this in [controller.go](https://github.com/aws-controllers-k8s/code-generator/blob/25c43e827527b43c652b6e1995265c8d6027567f/pkg/generate/ack/controller.go#L102) don't need to worry about Struct or what type.. just pass container. Leaves it open-ended for override types too.
```

Comment:
This was done in an effort to use these code generator methods within service controller template files. I still think that use case is valuable, and all of how we access and pass parameters to any of the `Set*` code generator methods needs some review with this in mind.

Reply:
Could you clarify a bit? Is there a reason why we can't use setForContainer vs. the concrete type?

Reply:
Initially every method was private, but we needed to expose some of these for custom hook code. I actually can't find where either of these are being used anymore (see [this search](https://github.com/search?q=org%3Aaws-controllers-k8s+SetSDKForStruct&type=code)), so I'm not too sure why we didn't expose the container. 

From memory, it was because the container version requires we pass an additional field that wasn't available within the hook code. But that's a vague guess at best.